### PR TITLE
clatd: add module and test

### PR DESCRIPTION
**Pull request by hax404**
## Description of changes
add module and test for clatd #304477

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/tests/clatd.nix [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+          address = [
+            "2001:db8::2/64"
+          ];
+          routes = [
+            { routeConfig = { Destination = "::/0"; Gateway = "2001:db8::1"; }; }
+          ];
+        };
+      };
+
+      systemd.services.clatd = {
```
Comment (jmbaur) [inline nixos/tests/clatd.nix:157]: Can we factor this out into a module? I don't think many options would be required, just an `enable` and `settings`, where `settings` can be translated to a config file for clatd using `pkgs.formats.keyValue`

**Review by jmbaur — APPROVED** [review]

**Comment by Majiir** [comment]
I've been trying this out. It generally works, but I get an error when my machine starts up:

```
systemd[1]: Started 464XLAT CLAT daemon.
clatd[1537]: Starting clatd v1.6 by Tore Anderson <tore@fud.no>
clatd[1537]: Using PLAT (NAT64) prefix: 64:ff9b::/96
clatd[1641]: RTNETLINK answers: Network is unreachable
clatd[1537]: <error> get_plat_dev(): 'ip -6 route get 64:ff9b::' failed
systemd[1]: clatd.service: Main process exited, code=exited, status=1/FAILURE
systemd[1]: clatd.service: Failed with result 'exit-code'.
```

Everything works fine after manually restarting the service. This machine has a straightforward SLAAC setup.

I see that you have `After=network-online.target`. Do you also need `Wants`? [This article](https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/) suggests it.

File: nixos/modules/services/networking/clatd.nix [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.clatd = {
+      description = "464XLAT CLAT daemon";
+      documentation = [ "man:clatd(8)" ];
+      wantedBy = [ "multi-user.target" ];
```
Comment (jmbaur) [inline nixos/modules/services/networking/clatd.nix:39]:
```suggestion
      wantedBy = [ "multi-user.target" "network-online.target" ];
```

File: nixos/modules/services/networking/clatd.nix [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      after = [ "network-online.target" ];
```
Comment (jmbaur) [inline nixos/modules/services/networking/clatd.nix:40]:
```suggestion
      after = [ "network-online.target" ];
      wants = [ "network-online.target" ];
```

File: nixos/modules/services/networking/clatd.nix [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/clatd -c ${configFile}";
+        StartLimitIntervalSec = 0;
+
+        # Hardenin
```
Comment (jmbaur) [inline nixos/modules/services/networking/clatd.nix:46]:
```suggestion
        # Hardening
```

**Comment by jmbaur** [comment]
> I've been trying this out. It generally works, but I get an error when my machine starts up:
> ...
> Everything works fine after manually restarting the service. This machine has a straightforward SLAAC setup with NetworkManager.

Is this due to the route not being available immediately on startup, but by the time you restart the service, the route now exists? The project recommends [here](https://github.com/toreanderson/clatd?tab=readme-ov-file#invocation) that clatd does not recognize changes in network connectivity and should be restarted by some external process. I've been using networkd-dispatcher for this, but it appears NetworkManager also has ways to hook into changes in connectivity.

File: nixos/modules/services/networking/clatd.nix [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  config = mkIf cfg.enable {
+    systemd.services.clatd = {
+      description = "464XLAT CLAT daemon";
+      documentation = [ "man:clatd(8)" ];
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network-online.target" ];
+
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/clatd -c ${configFile}";
+        StartLimitIntervalSec = 0;
```
Comment (jmbaur) [inline nixos/modules/services/networking/clatd.nix:44]: This needs to be under `unitConfig`, or we can use [this](https://search.nixos.org/options?channel=unstable&show=systemd.user.services.%3Cname%3E.startLimitIntervalSec&from=0&size=50&sort=relevance&type=packages&query=startlimit) option

**Review by jmbaur — APPROVED** [review]
Thanks @hax404 !

**Comment by hax404** [comment]
> > I've been trying this out. It generally works, but I get an error when my machine starts up:
> > ...
> > Everything works fine after manually restarting the service. This machine has a straightforward SLAAC setup with NetworkManager.
> 
> Is this due to the route not being available immediately on startup, but by the time you restart the service, the route now exists? The project recommends [here](https://github.com/toreanderson/clatd?tab=readme-ov-file#invocation) that clatd does not recognize changes in network connectivity and should be restarted by some external process. I've been using networkd-dispatcher for this, but it appears NetworkManager also has ways to hook into changes in connectivity.

I've put the script https://github.com/toreanderson/clatd/blob/master/scripts/clatd.networkmanager into `networking.networkmanager.dispatcherScripts` and it reliably restarts clatd when network changes.

**Review by nicoonoclaste — APPROVED** [review]

**Pull request merged**